lib/auth/autoupdate/autoupdatev1/service.go

```diff
@@ -648,6 +652,134 @@ func (s *Service) DeleteAutoUpdateAgentRollout(ctx context.Context, req *autoupd
+
+	var existingRollout, newRollout *autoupdate.AutoUpdateAgentRollout
+	for range maxTries {
+		existingRollout, err = s.backend.GetAutoUpdateAgentRollout(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		err = rollout.TriggerGroups(existingRollout, req.Groups, req.DesiredState, s.clock.Now())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		newRollout, err = s.backend.UpdateAutoUpdateAgentRollout(ctx, existingRollout)
+		if err == nil {
+			return newRollout, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return nil, trace.LimitExceeded("max update tries exceeded")
+}
+
+// ForceAutoUpdateAgentGroup forces one or many groups of the agent rollout into
+// a DONE state.
+func (s *Service) ForceAutoUpdateAgentGroup(ctx context.Context, req *autoupdate.ForceAutoUpdateAgentGroupRequest) (result *autoupdate.AutoUpdateAgentRollout, err error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindAutoUpdateAgentRollout, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminActionAllowReusedMFA(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	const maxTries = 3
+
+	var existingRollout, newRollout *autoupdate.AutoUpdateAgentRollout
+	for range maxTries {
+		existingRollout, err = s.backend.GetAutoUpdateAgentRollout(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		err = rollout.ForceGroupsDone(existingRollout, req.Groups, s.clock.Now())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		newRollout, err = s.backend.UpdateAutoUpdateAgentRollout(ctx, existingRollout)
+		if err == nil {
+			return newRollout, nil
+		}
+		if !trace.IsCompareFailed(err) {
+			return nil, trace.Wrap(err)
+		}
+	}
+	return nil, trace.LimitExceeded("max update tries exceeded")
+}
+
+// RollbackAutoUpdateAgentGroup forces one or many groups of the agent rollout into
+// a DONE state.
+func (s *Service) RollbackAutoUpdateAgentGroup(ctx context.Context, req *autoupdate.RollbackAutoUpdateAgentGroupRequest) (result *autoupdate.AutoUpdateAgentRollout, err error) {
+	authCtx, err := s.authorizer.Authorize(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.CheckAccessToKind(types.KindAutoUpdateAgentRollout, types.VerbUpdate); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := authCtx.AuthorizeAdminActionAllowReusedMFA(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	const maxTries = 3
+
+	var existingRollout, newRollout *autoupdate.AutoUpdateAgentRollout
+	for range maxTries {
+		existingRollout, err = s.backend.GetAutoUpdateAgentRollout(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		if len(req.Groups) == 0 {
+			err = rollout.RollbackStartedGroups(existingRollout, s.clock.Now())
+		} else {
+			err = rollout.RollbackGroups(existingRollout, req.Groups, s.clock.Now())
+		}
```

Comment:
This design makes it a bit harder than it should be to make changes to the RPC in the future, since dropping the "groups" field will result in potentially unwanted behavior. Can we have a boolean flag that signals the desire to rollback all the started groups instead, and an error if the groups are empty and the flag is unset?